src/permutation.jl

```diff
@@ -0,0 +1,162 @@
+         se = SVector{p,T}(stderror!(βsc, mod)),
+         θ = SVector{k,T}(getθ!(θsc, mod)),
+        )
+    end
+    MixedModelPermutation(
+        samp,
+        deepcopy(morig.λ),
+        getfield.(morig.reterms, :inds),
+        copy(morig.optsum.lowerbd),
+        NamedTuple{Symbol.(fnames(morig))}(map(t -> (t.cnames...,), morig.reterms)),
+    )
+end
+
+function permutation(nsamp::Integer, m::LinearMixedModel; kwargs...)
+    return permutation(Random.GLOBAL_RNG, nsamp, m; kwargs...)
+end
+
+function permutation(rng::AbstractRNG, n::Integer,
+                                morig::GeneralizedLinearMixedModel;
+                                use_threads::Bool=false) where {T}
+
+    throw(ArgumentError("GLMM support is not yet implemented"))
+end
+
+permute!(mod::LinearMixedModel, blups=ranef(mod), reterms=mod.reterms) =
+    permute!(GLOBAL_RNG, mod, blups, reterms)
+
+"""
+    permute!([rng::AbstractRNG,] mod::LinearMixedModel,
+              blups=ranef(mod), reterms=mod.reterms)
+
+Simulate and install a new response via sign-flipped permutation of the residuals
+at the observational and sign-flipping of the conditional modes at group level.
+
+Sign-flipped permutation of the residuals is similar to permuting the
+(fixed-effects) design matrix. Sign-flipping the conditional modes (random effects)
+preserves the correlation structure of the random effects, while also being equivalent to
+permutation via swapped labels for categorical variables.
+
+!!! note
+    This method has serious limitations for singular models because permuting zeros
+
+See also [`permutation`](@ref), [`nonparametricbootstrap`](@ref) and [`resample!`](@ref).
+
+# Reference
+ter Braak?
+
+"""
+function permute!(rng::AbstractRNG, mod::LinearMixedModel,
+                   blups=ranef(mod),
+                   reterms=mod.reterms)
+    β = coef(mod)
+    y = response(mod) # we are now modifying the model
+    copy!(y, residuals(mod))
+
+    shuffle!(rng, y)
+    # inflate these to be on the same scale as the empirical variation instead of the MLE
+    # y .*= sdest(mod) / std(y; corrected=false)
+    # sign flipping
+    y .*= rand(rng, (-1,1), length(y))
+
+    # σ = sdest(mod)
+
+    for (re, trm) in zip(blups, reterms)
+        npreds, ngrps = size(re)
+        samp = sample(rng, 1:ngrps, ngrps; replace=false)
+
+        # use a view to avoid copying
+        newre = view(re, :, samp)
+        # sign flipping
+        newre *= diagm(rand(rng, (-1,1), ngrps))
+
+        # inflation
+        #λmle =  trm.λ * σ                               # L_R in CGR
+        #λemp = cholesky(cov(newre'; corrected=false)).L # L_S in CGR
+        # no transpose because the RE are transposed relativ to CGR
+        #inflation = λmle / λemp
+
+        # our RE are actually already scaled, but this method (of unscaledre!)
+        # isn't dependent on the scaling (only the RNG methods are)
+        MixedModels.unscaledre!(y, trm, newre)
```

Comment:
add in the random effects to the observation-level residuals

Reply:
Here you do:
Zb* + r*
were the b* and r* is sampled random effect and residuals?

Reply:
yes